analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -18,16 +18,55 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks.CSharp;
 
 public class PublicMethodArgumentsShouldBeCheckedForNull : SymbolicRuleCheck
 {
     private const string DiagnosticId = "S3900";
-    private const string MessageFormat = "Refactor this method to add validation of parameter '{0}' before using it.";
+    private const string MessageFormat = "{0}";
 
     internal static readonly DiagnosticDescriptor S3900 = DescriptorFactory.Create(DiagnosticId, MessageFormat);
 
     protected override DiagnosticDescriptor Rule => S3900;
 
-    public override bool ShouldExecute() => false;
+    public override bool ShouldExecute() =>
+        Node is BaseMethodDeclarationSyntax or AccessorDeclarationSyntax;
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        var operation = context.Operation.Instance;
+        if (operation.Kind == OperationKindEx.ParameterReference
+            && operation.ToParameterReference().Parameter is var parameter
+            && parameter.Type.IsValueType is false
+            && IsParameterDereferenced(context.Operation)
+            && NullableStateIsNotKnownForParameter(parameter)
+            && !IgnoreBecauseOfParameterAttribute(parameter))
+        {
+            var message = SemanticModel.GetDeclaredSymbol(Node).IsConstructor()
+                ? "Refactor this constructor to avoid using members of parameter '{0}' because it could be null."
+                : "Refactor this method to add validation of parameter '{0}' before using it.";
+            ReportIssue(operation, string.Format(message, operation.Syntax), context);
+        }
+
+        return context.State;
+
+        bool NullableStateIsNotKnownForParameter(IParameterSymbol symbol) =>
+            !context.HasConstraint(symbol, ObjectConstraint.Null)
+            && !context.HasConstraint(symbol, ObjectConstraint.NotNull);
```

Comment:
Like this
```suggestion
            context.State[symbol] is null || !context.State[symbol].HasConstraint<ObjectConstraint>();
```
or add `HasConstraint<T>(ISymbol)` to the context. I haven't seen any other need for the extension, so maybe later.

There are other syntax ways how to do it, but with the `!`, they become ugly pretty quickly